### ISSUE TITLE
Extensions support

### ISSUE THREAD

**Issue by timsueberkrueb** [lirios/browser#22]
Support extensions, both declarative and script-based.

This was split up in the following milestones:

## 1. Base
Declarative:
* [x] Search engines
* [x] Themes

TODO:
* [x] Improve error handling (e.g. in `extensionparser.cpp`)
* [x] Static field naming (e.g. `search.query` in `searchprovider.cpp`
      Use `$[search.query]` or something like this instead to make clear it's a field not a value?
* [x] Fall back to `default` if other extension doesn't exist anymore
* [x] Search engines: support hash-bang urls like `https://peekier.com/#!search query`

## 2. Scripting
Dynamic:
* [ ] Omnibar plugins (omniplets)

## 3. GUI
For discussion:
* How should users create/search/install extensions?

**Comment by timsueberkrueb** [comment]
Work in progress on [feature/extensions](https://github.com/lirios/browser/tree/feature/extensions).

**Comment by timsueberkrueb** [comment]
This is now at a state where the base is implemented (declarative extension support). In order to prevent this branch growing to big, I think it is time to split this up. I will prepare to create a pull request to merge the base with develop and a new branch will be created for the next steps (which are script support and a better gui).

**Comment by timsueberkrueb** [comment]
New branch for extensions base:  [feature/extensions_base](https://github.com/lirios/browser/tree/feature/extensions_base).

**Comment by ghost** [comment]
In my mind, the extensions ecosystem should work as [npmjs](https://www.npmjs.com/), but with a few tweaks that would make it even better:

**Basic functionalities:**

- Hosting a kind of "marketplace"(though I would keep extensions free) on the main site( for example `liri.io/marketplace` or if you prefer `marketplace.liri.io`). The web system should have a decent user administration system and build process(see point 2)

- If the extensions need to be compiled, I would suggest doing it in the cloud, to avoid "compiler missing" errors or cross-platform issues, and to lower the download size(wich in some countries like Italy it's a real bummer)  and the installation time.
On the other hand, if you can't afford a solid server you should let the user download(from GitHub is the best I think) and the compile the extension itself. This should be a little bit slower, but you could have a nice progress bar and even some logging of the compiling going on. Let's say that the second option is more for `developer/advanced users`

- A page (for example `liri://extensions`) to manage all the extensions, uninstall them and disable/enable on startup; A welcome feature would be to load the extension right when the browser is opened(without re-launching the application)

If you're interested I could help on build system/web application.

**Comment by timsueberkrueb** [comment]
Hey @Luke123443, thanks for submitting your ideas and suggestions about this issue! Of course a proper market place for extensions would be brilliant, but I'm afraid that this is probably not feasible for different reasons - at least not at this point. 
What I thought about for a start is similar to the second part of your second point. Liri browser extensions are basically ZIP files, there's no real build process needed. Scripts will be JavaScripts, no need for building either. Thus, developers could host their extensions on GitHub as repositories. They tag their repo with the topic `lbx-repo` and possibly some additional tags like `lbx-theme`/`lbx-search` etc. to make them discoverable. Liri Browser would use the GitHub API and look for extension repositories by filtering by these tags. Whenever an extensions creator wants to release his extension, he can simply create a tag/GitHub release and Liri Browser will either look for an attached `.lbx` file *or* just fetch the `.zip` of that git tag. 
Point 3 is exactly what I'd like to see, too.
Anyway, those are just ideas, before we start deciding on this, we need to finish the base implementation (see #43).

> If you're interested I could help on build system/web application.

If we, however, decide that it is feasible and we can afford a proper market place, we'd be glad if you'd like to join the effort!

**Comment by ghost** [comment]
> What I thought about for a start is similar to the second part of your second point. Liri browser extensions are basically ZIP files, there's no real build process needed. Scripts will be JavaScripts, no need for building either. Thus, developers could host their extensions on GitHub as repositories. They tag their repo with the topic lbx-repo and possibly some additional tags like lbx-theme/lbx-search etc. to make them discoverable.
 
So you're basically planning to code only for the fontend until you have a solid back-end base(wich sounds perfect; I think that your idea is probably the best for now(since we don't need to compile the scripts, well obviously if we have some huge addons developers could use webpack or some bundling tool), but to make this compilation option of serving minified files(or simply remove extra weight) would be feasible through pulling the script data from a binary release on github,wich is a simple thing to do(as simple as cloning the repo) and you could then take advantage of git's version control to manage releases or even sync with develop branches. I don't know yet how you want to create this page,but i think that the way google does this is the best: they export an api to the window object of each webview, and the `liri://extensions` could be written in classic web languages,but event addons could be installed by external site with just one click(and even external site could create external extension manager to interact with the local extensions). In my opinion this would be an amazing feature, but the access to this api should be limited like notifications are in chrome

**Comment by timsueberkrueb** [comment]
@Luke123443 `liri://extensions` is already written using our native UI framework in QML. But there should be a way for extensions to expose custom settings which we could implement by allowing extensions to provide some HTML code we'd embed but I personally would prefer a more limited model which would ensure those settings to fit the general design.
Yep, we definitely want to have a permission system for script-based extensions.

**Comment by Aldrog** [comment]
@tim-sueberkrueb did you consider making scripting API compatible with Chrome's one?
https://developer.chrome.com/extensions/api_index

**Comment by timsueberkrueb** [comment]
@Aldrog That sounds pragmatic. We wouldn't cover everything that chrome supports though and our manifest files are different.
We could also try to completely support chrome extensions but that would be an enormous effort (also too keep up with changes) I fear.

**Comment by Aldrog** [comment]
Even partial support might be good to ease porting. It would be cool if you could take a Chrome extension, write an lbx manifest, replace all chrome.* calls with liri.* and the extension would work in most cases.

Also, designing a new API is a hard task, embracing an existing one is easier.

**Comment by timsueberkrueb** [comment]
We decided to lay down extensions support to concentrate on more important features for now.
